### ISSUE TITLE
Need item details for error "40002 - Item quantity unavailable"

### ISSUE THREAD

**Issue by nirmay** [Open-network-for-digital-commerce/ONDC-Protocol-Specs#37]
Just returning the error with message "Item quantity unavailable" is not sufficient for BAP.
In case, there are many items added to the cart, BAP will need to present to the user, which item(s) are low on quantity.

I would like to propose, along with error, we pass following

message.order.items (which is an array) and BPP sends back current quantity as per below,

`{
                    "id": "73661421",
                    "quantity": {
                        "count": 0
                    }
                },`

@BLR-0118 pls review,

**Comment by EkanshKr** [comment]
This should work. 
We should keep the quote.breakup too in the message.order. For example - There are 4 products in the /init call and for 1 product we may get quantity mismatch and for another one we may get price update error. It's better to cover these two scenarios in one /on_init request (quantity unavailable and price update can happen for the same product too). 
As we have “@ondc/org/item-qty” in the quote.breakup object now, so BPPs can send the price according to available quantity and can mention the qty in this field too.
@BLR-0118  @nirmay  any challenges if we add quote in this?

**Comment by mahoriR** [comment]
@nirmay This is good. 

So if i have to generalize this - 

In case of this error - on_init the response will have the actual available quantities (less then or equal to what is asked In init).
So if requestor asks for 2 items , 3 quantity each , we could respond with - 
[{ "id": "73661421", "quantity": { "count": 1 } },{ "id": "73661421", "quantity": { "count": 3 } },]
and an error for insufficient quantity.

This is right ?

**Comment by mahoriR** [comment]
@EkanshKr While the order object can have multiple changes (prices, quantities) but error object today doesn't have a way to convey multiple errors. That would be a problem right.
 @BLR-0118

**Comment by nirmay** [comment]
> @nirmay This is good.
> 
> So if i have to generalize this -
> 
> In case of this error - on_init the response will have the actual available quantities (less then or equal to what is asked In init). So if requestor asks for 2 items , 3 quantity each , we could respond with - [{ "id": "73661421", "quantity": { "count": 1 } },{ "id": "73661421", "quantity": { "count": 3 } },] and an error for insufficient quantity.
> 
> This is right ?

In my proposal, the items array should include only those items that fail the inventory quantity check.
if buyer orders 10 items and 2 of them fail the check then only those are returned in the on_init response inside the items array.

**Comment by VidyaGF** [comment]
Does this cover if we have two error IDs , quantity error and price error , we will need to display multiple error messages for 2 different items in the same checkout page.

**Comment by nirmay** [comment]
> Does this cover if we have two error IDs , quantity error and price error , we will need to display multiple error messages for 2 different items in the same checkout page.

good point, suggest you file a separate bug to address that.

**Comment by EkanshKr** [comment]
I am not sure but currently i don't think that we get any price update error object. Buyer app checks the price at their end and the price what we received in breakup object of  /on_init response. do we need different error code for price update? 
By Adding breakup object also in the case of quantity error , buyer app can check the price update and quantity in one /on_init response. 
@nirmay @VidyaGF @mahoriR

**Comment by VidyaGF** [comment]
I was talking more of making the error object generic so that we can handle multiple errors and item-wise , so that anytime in future, different types of errors can be reported. Buyer app can read list of errors and make provision to display  it appropriately to the user.

**Comment by BLR-0118** [comment]
@nirmay @mahoriR - agree that we can use this approach only to indicate inventory quantity check failing;
@EkanshKr - price update by seller - is that an error?
Suggest we go ahead with the original reason for which this issue was filed i.e. inventory quantity check failing. Not sure how we can handle multiple errors in the same error object?

**Comment by EkanshKr** [comment]
@nirmay @BLR-0118 @VidyaGF @mahoriR 
sharing the on_init response for this issue as discussed with Supriyo and Mayuresh - 

https://docs.google.com/document/d/19dHfFegVwWIYdnVvE8uUXPZsTl9MjOLjsUv8TAXU5Yw/edit?usp=sharing

**Comment by EkanshKr** [comment]
@BLR-0118  Yes Supriyo , we have generic error handler to show these messages and we are showing the other error messages like "delivery agent not available" , "price has been updated" through that handler only.

**Comment by BLR-0118** [comment]
If > 1 items in error, send an error code such as "mismatch between what is selected and what is available" and for Paytm / other buyer apps to highlight items where there is an issue. If this works, I can define the error code.
@EkanshKr @mahoriR @nirmay @VidyaGF @akil4n
